lib/core_to_mucore.ml

```diff
@@ -1052,19 +1094,32 @@ let make_function_args f_i loc env args (accesses, requires) =
       (*   (LTranslate.T (IT.good_ (ct, IT.sym_ (pure_arg, bt, here)) here), info) *)
       (* in *)
       let@ at =
-        aux (arg_states @ [ (mut_arg, arg_state) ]) (* good_lc :: *) good_lcs env st rest
+        aux
+          (arg_states @ [ (mut_arg, arg_state) ])
+          (* good_lc :: *) good_lcs
+          env
+          st
+          (rest, ghost_args)
       in
       return (Mu.mComputational ((pure_arg, bt), (loc, None)) at)
-    | [] ->
-      let@ lat = make_largs_with_accesses (f_i arg_states) env st (accesses, requires) in
+    | [], (arg, cnbt) :: rest ->
+      let sbt = Compile.base_type env cnbt in
+      let bt = SBT.proj sbt in
+      let env = Translate.add_logical arg sbt env in
+      let@ at = aux arg_states good_lcs env st ([], rest) in
+      return (Mu.Ghost ((arg, bt), (loc, None), at))
+    | [], [] ->
+      let@ lat =
+        make_largs_with_accesses (f_i arg_states) env st (accesses, snd requires)
+      in
       return (Mu.L (Mu.mConstraints (List.rev good_lcs) lat))
   in
-  aux [] [] env Translate.C_vars.init args
+  aux [] [] env Translate.C_vars.init (args, fst requires)
```

Comment:
Would this be clearer as two recursive functions, rather than one with `ghost_args` threaded through?